### PR TITLE
BIO-633 - Add support for variant "dup" type

### PR DESCRIPTION
**Pull request by jhl667**
Added section for "dup" variants in correct_indel_coords.  This requires a class for pysamFasta since we need to find the reference sequence to determine whether our alternate allele is in fact a dup.   Might want to take a look at the pysam_file object since it is being passed around quite a bit before it makes it to correct_indel_coords.

### CODE REVIEW

**Review by jPleyte — COMMENTED** [review]
I rebased this branch and ended up putting it in a different one instead of rewriting this one. I suggest this pull request be closed w/o merging.

File: cgd_tx_eff/src/edu/ohsu/compbio/txeff/tx_eff_hgvs.py [hunk 1]
```diff
@@ -72,10 +102,20 @@ def _correct_indel_coords(pos, ref, alt):
...
+                new_pos = '_'.join([str(new_start), str(new_end)]) + 'dup'
         # Insertion case
-        new_start = str(pos)
-        new_end = str(int(pos) + 1)
-        new_pos = '_'.join([new_start, new_end]) + 'ins' + alt[1:]
+        else:
+            new_start = str(pos)
+            new_end = str(int(pos) + 1)
+            new_pos = '_'.join([new_start, new_end]) + 'ins' + alt[1:]
+        print(new_pos)
```
[the remark this answers is not in the document]
Comment (jhl667) [reply cgd_tx_eff/src/edu/ohsu/compbio/txeff/tx_eff_hgvs.py]: ah, yes that was supposed to be removed before commit

File: cgd_tx_eff/tool-data/all_fasta.loc.sample
[the remark this answers is not in the document]
Comment (jhl667) [reply cgd_tx_eff/tool-data/all_fasta.loc.sample:1]: yes, these are required by Galaxy so that you can access reference genomes on the filesystem.  The all_fasta loc file will contain an entry that names the genome (eg hg19) and the actual location.  The reason it does this is that several index files are created for reference genome, which would require some sort of linking on Galaxy's end between the various indices and the actual FASTA.  Instead, it just uses this other system to point directly at them.

File: cgd_tx_eff/src/edu/ohsu/compbio/txeff/tx_eff_control.py [hunk 2]
```diff
@@ -64,9 +70,14 @@ def _main():
     # Use tx_eff_annovar to read annovar records 
     annovar_records = tx_eff_annovar.get_annovar_records(args.annovar_variant_function.name, args.annovar_exonic_variant_function.name)
 
+    # Load the reference genome with pysam.
+    pysam_file = PysamTxEff(args.reference_fasta)
+
     # Use tx_eff_hgvs to fix the nomenclature
     tx_eff_hgvs.identify_hgvs_datasources() 
-    merged_transcripts = tx_eff_hgvs.get_updated_hgvs_transcripts(annovar_records)
+    merged_transcripts = tx_eff_hgvs.get_updated_hgvs_transcripts(annovar_records, pysam_file)
```
[the remark this answers is not in the document]
Comment (jhl667) [reply cgd_tx_eff/src/edu/ohsu/compbio/txeff/tx_eff_control.py:77]: I just meant that I was having to call this pysam object in multiple places, seemed like there was probably a cleaner way to do it.

File: cgd_tx_eff/src/edu/ohsu/compbio/txeff/tx_eff_hgvs.py [hunk 3]
```diff
@@ -45,7 +46,36 @@ def _noneIfEmpty(value: str):
...
+
+class DupHandler:
+    def __init__(self, pysam_file: PysamTxEff, chrom: str, coord: int, alt: str):
+        self.pysam_file = pysam_file
+        self.chrom = chrom
+        self.coord = coord
+        self.alt = alt[1:]
+        self.size = self.pysam_file.size
+        self.seq = self.pysam_file.faidx_query(self.chrom, self.coord)
+        self.fasta_alt = self.seq[self.size + 1:self.size + len(self.alt) + 1]
```
[the remark this answers is not in the document]
Comment (jhl667) [reply cgd_tx_eff/src/edu/ohsu/compbio/txeff/tx_eff_hgvs.py:58]: - self.alt is defining the alt allele sequence that we are using to search for a duplicate (from the VCF)
- self.size is a constant defined to limit the amount of sequence we look at in the reference FASTA
- self.seq is the actual sequence from the reference, defined based on the position of the alt allele and the constant self.size
- self.fasta_alt provides the sequence directly following the alternate allele, it is trying to match them to see whether the sequence is duplicated

Basically the idea is the move down the reference sequence and keep looking to see whether the given alternate allele is duplicated.  So, if there is an ATGT insertion, this would query the four bases immediately following the ATGT insertion and find out if we once again saw an ATGT.  If so, that would be a duplication.

File: cgd_tx_eff/src/edu/ohsu/compbio/txeff/tx_eff_hgvs.py [hunk 4]
```diff
@@ -45,7 +46,36 @@ def _noneIfEmpty(value: str):
         return None
     return value
 
-def _correct_indel_coords(pos, ref, alt):
+
+class DupHandler:
+    def __init__(self, pysam_file: PysamTxEff, chrom: str, coord: int, alt: str):
```
[the remark this answers is not in the document]
Comment (jhl667) [reply cgd_tx_eff/src/edu/ohsu/compbio/txeff/tx_eff_hgvs.py:51]: You can change it to something better, but I wouldn't use "pysam", as that is the name of the Python libraries this functionality is being provided from.

File: cgd_tx_eff/src/edu/ohsu/compbio/txeff/tx_eff_hgvs.py [hunk 5]
```diff
@@ -45,7 +46,36 @@ def _noneIfEmpty(value: str):
...
+            fasta_alt = self.seq[start_size:end_size]
+        new_start = coord - len(self.alt)
+        new_end = new_start + len(self.alt) - 1
+        return new_start, new_end
+
+
+def _correct_indel_coords(chrom, pos, ref, alt, pysam_file):
     """
     Using a VCF position, create coords that are compatible with HGVS nomenclature.
     Since we are already determining at this stage whether the event is an ins or del, also
```
[the remark this answers is not in the document]
Comment (jhl667) [reply cgd_tx_eff/src/edu/ohsu/compbio/txeff/tx_eff_hgvs.py:81]: latest HGVS standards specify you don't list the deleted allele, just the position

File: cgd_tx_eff/tool_data_table_conf.xml.sample
[the remark this answers is not in the document]
Comment (jhl667) [reply cgd_tx_eff/tool_data_table_conf.xml.sample:1]: see above, yes it is needed

**Comment by jPleyte** [comment]
To Do: Delete this branch once we're confident the changes were merged to master and tested.
